Update resource reservation end date

victor-champonnois opened the pull request:
[Task](https://gestion.coopiteasy.be/web#id=7527&view_type=form&model=project.task&action=475&active_id=198)

[Related PR](https://github.com/coopiteasy/cie-custom/pull/47)

- create set_activity_as_done method that updates the state and the reservation end date
- call this method instead of just updating the state

victor-champonnois committed:
[IMP] Update resource reservation end date

- create set_activity_as_done method that updates the state and the reservation end date
- call this method instead of just updating the state

robinkeunen commented on resource_activity/models/resource_activity.py:
```diff
@@ -368,12 +368,17 @@ def action_done(self):
+    def set_activity_as_done(self):
+        self.state = "done"
+        if self.date_end > datetime.now():
+            self.registrations.allocations.date_end = datetime.now()
```
I think this won't work if there is more than one registration on the activity or more than one allocation on the registration.

robinkeunen replied:
Also, i'd add `api.multi` and `self.ensure_one()`

victor-champonnois replied:
Fixed, I also refactored to simplify the tests

robinkeunen reviewed (changes requested):
Can you add some unit tests as well ?

victor-champonnois committed:
[ADD] unit test update_resource_registration_date_end

victor-champonnois commented:
- added unit tests
- refactored to simplify testing (avoid depending on datetime.now after reading [this](https://hakibenita.com/python-dependency-injection))
- adding support for multiple registrations/allocations
- adding `@api.multi` and `ensure_one()`

robinkeunen commented on resource_activity/tests/test_resource_activity.py:
```diff
@@ -114,3 +114,93 @@ def test_activity_w_booked_resources(self):
+        )
+        activity_not_finished.search_all_resources()
+        activity_not_finished.reserve_needed_resource()
+        activity_not_finished.update_resource_registration_date_end(date_done)
```
Shouldn't you rather close the activity and test that `set_activity_as_done` and `update_resource_registration_date_end` were called ?

victor-champonnois replied:
Yes you are right, see what I did in my answer below

robinkeunen reviewed (changes requested):
Big Thank you for the unit tests. I'd go a step further and test each case in a separate unit test.

Also, very interesting note about date injection.

```python
class TestResourceActivity(test_base.TestResourceActivityBase):
    def setUp(self):
        super().setUp()
        # prepare common setup data 
        date_start = datetime(2022, 2, 1, 1, 1, 1)
        registration_1 = {
            "attendee_id": self.partner_demo.id,
            "quantity": 1,
            "quantity_needed": 1,
            "booking_type": "booked",
            "resource_category": self.mtb_category.id,
            "product_id": self.bike_product.id,
        }
        ...

    def test_activity_ended_early_updates_registrations_date_end(self):
    
        date_start = datetime(2022, 2, 1, 1, 1, 1)
        date_end = date_start + timedelta(days=4)
        date_done = date_start + timedelta(days=2)
        ...
        
        for registration_date_end in activity_not_finished.registrations.mapped(
            "allocations").mapped("date_end"):
            self.assertEqual(registration_date_end, date_done)
        activity_not_finished.unreserve_resources()
        
    
    def test_activity_set_as_done_after_end_does_not_updates_registrations_date_end
        date_end = date_start + timedelta(days=4)
        date_done = date_start + timedelta(days=6)
        ...
        for registration_date_end in activity_already_finished.registrations.mapped("allocations").mapped("date_end"):
            self.assertEqual(registration_date_end, date_end)
```


You'll see that I use lengthy descriptive names for the functions : they only exixt here.

Some day, I would like to implement [these best practices as well](https://www.youtube.com/watch?v=mks7haGqmKY) but I have yet to get familiar w/ them.

robinkeunen commented:
I see my comment in the code breaks your suggestion to use dependency injection. Not sure what's the best way to go.

victor-champonnois commented:
- refactored the tests
- added tests to check that `action_done` changes the state and changes the date_end. 
    - Because of the use of datetime_now() in action_done, I could not precisely test for the date change so I just tested that `datetime.now()` is greater than the updated date (assertAlmostEqual does not seem to work with datetime). The precise check is done with the first test. 
    - Also did not succeed to use `mock` in order to mock `datetime` (odoo modules might not behave as normal python package ?)
    - We could consider using `freezegun` as explained in the video

robinkeunen reviewed (approved):
Super travail => it'd be nice if you could share your findings about date injection at the CC.

victor-champonnois commented:
Using freezegun for tests

victor-champonnois committed:
[ADD] freezegun in test-requirements
